adapters/slog/slog.go

```diff
@@ -0,0 +1,256 @@
+// following environment variables:
+//
+//   - AXIOM_TOKEN
+//   - AXIOM_ORG_ID (only when using a personal token)
+//   - AXIOM_DATASET
+//
+// The configuration can be set manually using options which are prefixed with
+// "Set".
+//
+// An api token with "ingest" permission is sufficient enough.
+//
+// A handler needs to be closed properly to make sure all logs are sent by
+// calling [Handler.Close].
+func New(options ...Option) (*Handler, error) {
+	root := &rootHandler{
+		eventCh: make(chan axiom.Event, defaultBatchSize),
+		closeCh: make(chan struct{}),
+	}
+
+	handler := &Handler{
+		rootHandler: root,
+	}
+
+	// Apply supplied options.
+	for _, option := range options {
+		if err := option(handler); err != nil {
+			return nil, err
+		}
+	}
+
+	// Create client, if not set.
+	if root.client == nil {
+		var err error
+		if root.client, err = axiom.NewClient(root.clientOptions...); err != nil {
+			return nil, err
+		}
+	}
+
+	// When the dataset name is not set, use "AXIOM_DATASET".
+	if root.datasetName == "" {
+		root.datasetName = os.Getenv("AXIOM_DATASET")
+		if root.datasetName == "" {
+			return nil, ErrMissingDatasetName
+		}
+	}
+
+	// Run background ingest.
+	go func() {
+		defer close(root.closeCh)
+
+		logger := log.New(os.Stderr, "[AXIOM|SLOG]", 0)
+
+		res, err := root.client.IngestChannel(context.Background(), root.datasetName, root.eventCh, root.ingestOptions...)
+		if err != nil {
+			logger.Printf("failed to ingest events: %s\n", err)
+		} else if res.Failed > 0 {
+			// Best effort on notifying the user about the ingest failure.
+			logger.Printf("event at %s failed to ingest: %s\n",
+				res.Failures[0].Timestamp, res.Failures[0].Error)
+		}
+	}()
+
+	return handler, nil
+}
+
+// Close the handler and make sure all events are flushed. Closing the handler
+// renders it unusable for further use.
+func (h *Handler) Close() {
+	h.closeOnce.Do(func() {
+		close(h.eventCh)
+		<-h.closeCh
+	})
+}
+
+// Enabled implements [slog.Handler].
+func (h *Handler) Enabled(_ context.Context, level slog.Level) bool {
+	minLevel := slog.LevelInfo
+	if h.level != nil {
+		minLevel = h.level.Level()
+	}
+	return level >= minLevel
+}
+
+// Handle implements [slog.Handler].
+func (h *Handler) Handle(_ context.Context, r slog.Record) error {
+	event := axiom.Event{}
+
+	// Set record attributes first, handler attributes second.
+	r.Attrs(func(attr slog.Attr) bool {
+		addAttrToEvent(event, attr)
+		return true
+	})
+	for _, attr := range h.attrs {
+		addAttrToEvent(event, attr)
+	}
+
+	// Nest attributes in handler groups, if any.
+	for i := len(h.groups) - 1; i >= 0; i-- {
+		event = axiom.Event{h.groups[i]: event}
+	}
+
+	// Set timestamp, level and actual message. The zero time is ignored.
+	if !r.Time.IsZero() {
+		event[ingest.TimestampField] = r.Time.Format(time.RFC3339Nano)
+	}
+	event[slog.LevelKey] = r.Level.String()
+	event[slog.MessageKey] = r.Message
+
+	h.eventCh <- event
```

Comment:
If the channel is closed this will panic, can we not panic?

Reply:
Good call, changed this for all affected adapters.